server/stats/division_service.py

```diff
@@ -0,0 +1,179 @@
+        self._divisions = None
+        self._players = None
+        self.season = season
+        self.accessor = accessor
+
+    async def get_divisions(self) -> List['Division']:
+        if self._divisions is None:
+            self._divisions = await self.accessor.get_divisions()
+
+        return self._divisions
+
+    async def _ensure_players(self):
+        if self._players is None:
+            players_infos = await self.accessor.get_player_infos(self.season)
+            self._players = dict()
+
+            for info in players_infos:
+                self._players[info.user_id] = info
+
+    async def _get_players(self) -> Dict[int, 'PlayerDivisionInfo']:
+        await self._ensure_players()
+
+        return self._players
+
+    async def get_player(self, user_id: int):
+        return (await self._get_players())[user_id]
+
+    async def add_player(self, player_id: int) -> None:
+        await self._ensure_players()
+
+        logger.info("Added new player %s to divisions", player_id)
+        self._players[player_id] = PlayerDivisionInfo(player_id, 1, 0.0)
+        await self.accessor.add_player(self.season, self._players[player_id])
+
+    async def update_player_stats(self, player: PlayerDivisionInfo, new_score: float) -> None:
+        logger.debug("Update score for %s to %s", player)
+        player.score = new_score
+        await self.accessor.update_player(self.season, player)
+
+    async def promote_player(self, player):
+        logger.info("%s got promoted to league %s", player, player.league + 1)
+        player.score = 0.0
+        player.league += 1
+        await self.accessor.update_player(self.season, player)
+
+    async def post_result(self, player_one: int, player_two: int, winning_slot: int) -> None:
+        """
+        Post a ladder game result to the division system
+        :param player_one: FAF User ID of 1st player
+        :param player_two: FAF User ID of 2nd player
+        :param winning_slot: 0 for draw, 1 for 1st player, 2 for 2nd player
+        """
+        players = await self._get_players()
+
+        if player_one not in players:
+            await self.add_player(player_one)
+
+        if player_two not in players:
+            await self.add_player(player_two)
+
+        if winning_slot == 0:
+            logger.info("Game ended in a draw - no changes in score")
+            await self.update_player_stats(self._players[player_one], players[player_one].score)
+            await self.update_player_stats(self._players[player_two], players[player_two].score)
+            return
+
+        winner = players[player_one] if winning_slot == 1 else players[player_two]
+        loser = players[player_two] if winning_slot == 1 else players[player_one]
+
+        if winner.is_in_inferior_league(loser):
+            gain = 1.5
+            loss = 1.0
+        elif winner.is_in_superior_league(loser):
+            gain = 0.5
+            loss = 0.5
```

Comment:
I see 2 arguments against loss = 0.0:
Losing no points increases the chance that the player simply disconnects or Ctrl+Ks because he doesn't want to play against a more difficult enemy.
If you want to advance into a higher league, you need to be able to compete with them anyway. What's the point of getting into a league of players where you have no chance to defeat them?